engines/sherlock/animation.cpp

```diff
@@ -0,0 +1,208 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "sherlock/animation.h"
+#include "sherlock/sherlock.h"
+#include "common/algorithm.h"
+
+namespace Sherlock {
+
+static const int NO_FRAMES = FRAMES_END;
+
+Animation::Animation(SherlockEngine *vm): _vm(vm) {
+}
+
+/**
+ * Play a full-screen animation
+ */
+bool Animation::play(const Common::String &filename, int minDelay, int fade,
+		bool setPalette, int speed) {
+	Events &events = *_vm->_events;
+	Screen &screen = *_vm->_screen;
+	Sound &sound = *_vm->_sound;
+	int soundNumber = 0;
+	sound._playingEpilogue = true;
+
+	// Check for any any sound frames for the given animation
+	const int *soundFrames = checkForSoundFrames(filename);
+
+	// Add on the VDX extension
+	Common::String vdxName = filename + ".vdx";
+
+	// Load the animation
+	Common::SeekableReadStream *stream;
+	if (!_vm->_titleOverride.empty())
+		stream = _vm->_res->load(vdxName, _vm->_titleOverride);
+	else if (_vm->_useEpilogue2)
+		stream = _vm->_res->load(vdxName, "epilog2.lib");
+	else
+		stream = _vm->_res->load(vdxName, "epilogue.lib");
+
+	// Load initial image
+	Common::String vdaName = filename + ".vda";
+	ImageFile images(vdaName, true, true);
+
+	events.wait(minDelay);
+	if (fade != 0 && fade != 255)
+		screen.fadeToBlack();
+
+	if (setPalette) {
+		if (fade != 255)
+			screen.setPalette(images._palette);
+	}
+
+	int frameNumber = 0;
+	Common::Point pt;
+	bool skipped = false;
+	while (!_vm->shouldQuit()) {
+		// Get the next sprite to display
+		int imageFrame = stream->readSint16LE();
+
+		if (imageFrame == -2) {
+			// End of animation reached
+			break;
+		} else if (imageFrame != -1) {
+			// Read position from either animation stream or the sprite frame itself
+			if (imageFrame < 0) {
+				imageFrame += 32768;
```

Comment:
I think this is a matter of preference, but using "if (imageFrame & 0x8000)" and "imageFrame &= 0x7FFF" might make it more clear that the top bit decides on which case to use.